subsys/settings/src/settings.c

```diff
@@ -32,27 +31,113 @@ void settings_init(void)
 
 int settings_register(struct settings_handler *handler)
 {
-	if (settings_handler_lookup(handler->name)) {
-		return -EEXIST;
+	int rc;
+	struct settings_handler *ch;
+
+	k_mutex_lock(&settings_lock, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&settings_handlers, ch, node) {
+		/* generate the list in reverse alphabetical order, this will
+		 * guarantee the correct order of processing when subtrees
+		 * are registered. An example list should look like:
+		 *   bt/mesh/ttt/aaa
+		 *   bt/mesh/ttt
+		 *   bt/mesh
+		 *   bt
+		 * each of these can be registered independently.
+		 */
+		if (strcmp(handler->name, ch->name) < 0) {
+			break;
+		}
+		if (strcmp(handler->name, ch->name) == 0) {
+			rc = -EEXIST;
+			goto end;
+		}
+	}
+
+	sys_slist_insert(&settings_handlers, &ch->node, &handler->node);
+	rc = 0;
+end:
+	k_mutex_unlock(&settings_lock);
+	return rc;
+}
+
+int settings_deregister(struct settings_handler *handler)
+{
+	bool found;
+
+	k_mutex_lock(&settings_lock, K_FOREVER);
+	found = sys_slist_find_and_remove(&settings_handlers, &handler->node);
+	k_mutex_unlock(&settings_lock);
+
+	return found ? 0 : -EINVAL;
+}
+
+int settings_name_cmp(const char *name, const char *key, const char **next)
+{
+	if (next) {
+		*next = NULL;
+	}
+
+	/* name might come from flash directly, in flash the name would end
+	 * with '=' or '\0' depending how storage is done. Flash reading is
+	 * limited to what can be read
+	 */
+
+	while ((*key != '\0') && (*key == *name) &&
+	       (*name != '\0') && (*name != SETTINGS_NAME_END)) {
+		key++;
+		name++;
+	}
+
+	if (*key != '\0') {
+		return 0;
+	}
+
+	if (*name == SETTINGS_NAME_SEPARATOR) {
+		if (next) {
+			*next = name + 1;
+		}
+		return 1;
+	}
+
+	if ((*name == SETTINGS_NAME_END) || (*name == '\0')) {
+		return 1;
 	}
-	sys_slist_prepend(&settings_handlers, &handler->node);
 
 	return 0;
 }
 
-/*
- * Find settings_handler based on name.
- */
-static struct settings_handler *settings_handler_lookup(char *name)
+int settings_name_split(const char *name, char *argv, const char **next)
 {
-	struct settings_handler *ch;
+	if (next) {
+		*next = NULL;
+	}
 
-	SYS_SLIST_FOR_EACH_CONTAINER(&settings_handlers, ch, node) {
-		if (!strcmp(name, ch->name)) {
-			return ch;
+	/* name might come from flash directly, in flash the name would end
+	 * with '=' or '\0' depending how storage is done. Flash reading is
+	 * limited to what can be read
+	 */
+	while ((*name != '\0') && (*name != SETTINGS_NAME_END) &&
```

Comment:
This may be good point to use ```strspn```. This would make it also easier to configure.

Reply:
*only if can be used without newlib C.

Reply:
Ok... Sadly it seems that this is not a part of the minimal implementation.